Content/Items/Vitric/Weapons.CoachGunUpgrade.cs

```diff
@@ -0,0 +1,686 @@
+        }
+
+        public override Vector2? HoldoutOffset()
+        {
+            return new Vector2(-15, 0);
+        }
+
+        public override bool CanUseItem(Player player)
+        {
+            if (player.altFunctionUse == 2)
+            {
+                Item.useStyle = ItemUseStyleID.Swing;
+                Item.noUseGraphic = true;
+
+                Item.useTime = Item.useAnimation = 15;
+                if (player.ownedProjectileCounts[ModContent.ProjectileType<CoachGunUpgradeBomb>()] >= 3 || cooldown > 0)
+                    return false;
+            }
+            else
+            {
+                Item.useStyle = ItemUseStyleID.Shoot;
+                Item.noUseGraphic = false;
+
+                Item.useTime = Item.useAnimation = 30;
+            }
+            return base.CanUseItem(player);
+        }
+
+        public override void ModifyShootStats(Player player, ref Vector2 position, ref Vector2 velocity, ref int type, ref int damage, ref float knockback)
+        {
+            if (player.altFunctionUse == 2)
+            {
+                velocity = velocity * 0.8f;
+                type = ModContent.ProjectileType<CoachGunUpgradeBomb>();
+            }
+        }
+
+        public override bool Shoot(Player player, EntitySource_ItemUse_WithAmmo source, Vector2 position, Vector2 velocity, int type, int damage, float knockback)
+        {
+            if (player.altFunctionUse == 2)
+                cooldown = 30;
+            else
+            {
+                Vector2 offset = Vector2.Normalize(velocity) * 25f;
+
+                for (int k = 0; k < 12; k++)
+                {
+                    Dust.NewDustPerfect(position + offset * 1.5f, ModContent.DustType<Dusts.Glow>(), (velocity.RotatedByRandom(0.35f)) * Main.rand.NextFloat(0.15f, 0.45f), 70, Color.DarkOrange, Main.rand.NextFloat(0.2f, 0.5f));
+                }
+                Vector2 smokeVelo = new Vector2(1, -0.05f * player.direction).RotatedBy(velocity.ToRotation());
+                Dust.NewDustPerfect(position + offset * 1.5f, ModContent.DustType<Dusts.Smoke>(), Vector2.UnitY * -2 + smokeVelo.RotatedByRandom(0.35f) * 4.5f, 0, new Color(60, 55, 50) * 0.5f, Main.rand.NextFloat(0.5f, 1));
+
+                Helper.PlayPitched("Guns/PlinkLever", 0.45f, Main.rand.NextFloat(-0.1f, 0.1f), position);
+                Helper.PlayPitched("Guns/RifleLight", 0.75f, Main.rand.NextFloat(-0.1f, 0.1f), position);
+                Projectile.NewProjectileDirect(source, position, velocity * 1.4f, type, damage, knockback, player.whoAmI).GetGlobalProjectile<CoachGunUpgradeGlobalProj>().ShotFromGun = true;
+
+                Gore.NewGore(source, player.Center + (offset / 2), new Vector2(player.direction * -1, -0.5f) * 2, Mod.Find<ModGore>("CoachGunUpgradeCasing").Type, 1f);
+                return false;
+            }
+            return true;
+        }
+    }
+
+    public class CoachGunUpgradeGlobalProj : GlobalProjectile
```

Comment:
A full GlobalProjectile for a single flag like this feels wrong. Are there alternative designs to consider? Potentially consider a single ghost projectile to track being able to break the crystal? Or simply verifying owner only? It is extremely unlikely someone will be swapping guns fast enough to shoot the crystal with another gun.